webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreSerializer.scala

```diff
@@ -0,0 +1,114 @@
+
+  implicit val scoreValueEncoder: JsonEncoder[ScoreValue] = JsonEncoder[String].contramap(_.value)
+  implicit val scoreValueDecoder: JsonDecoder[ScoreValue] = JsonDecoder[String].mapOrFail(ScoreValue.fromString)
+
+  implicit val noDetailsScoreEncoder: JsonEncoder[NoDetailsScore] = DeriveJsonEncoder.gen
+  implicit val noDetailsScoreDecoder: JsonDecoder[NoDetailsScore] = DeriveJsonDecoder.gen
+
+  implicit val globalScoreEncoder: JsonEncoder[GlobalScore] = DeriveJsonEncoder.gen
+  implicit val globalScoreDecoder: JsonDecoder[GlobalScore] = DeriveJsonDecoder.gen
+
+  implicit val jsonScoreEncoder: JsonEncoder[JsonScore] = DeriveJsonEncoder.gen
+  implicit val jsonScoreDecoder: JsonDecoder[JsonScore] = DeriveJsonDecoder.gen
+}
+
+trait ScoreSerializerService {
+  def parse(score:  Score[Json]): IOResult[Option[Score[_]]]
+  def toJson(score: Score[_]):    IOResult[Option[Score[Json]]]
+}
+
+case object ComplianceScoreSerializerService extends ScoreSerializerService {
+  implicit val compliancePercentDecoder: JsonDecoder[ComplianceSerializable] = DeriveJsonDecoder.gen
+  implicit val compliancePercentEncoder: JsonEncoder[ComplianceSerializable] = DeriveJsonEncoder.gen
+
+  override def parse(score: Score[Json]): IOResult[Option[ComplianceScore]] = {
+
+    if (score.name == "compliance") {
+      score.details.as.toIO.map(ComplianceScore(score.value, score.message, _)).map(Some(_))
+    } else {
+      None.succeed
+    }
+  }
+
+  override def toJson(score: Score[_]): IOResult[Option[Score[Json]]] = {
+    score.details match {
+      case details: ComplianceSerializable =>
+        details.toJsonAST.toIO.map(JsonScore(score.value, score.name, score.message, _)).map(Some(_))
+      case _ => None.succeed
+    }
+  }
+}
+
+case object SystemUpdateScoreSerializerService extends ScoreSerializerService {
+  implicit val compliancePercentDecoder: JsonDecoder[SystemUpdateStats] = DeriveJsonDecoder.gen
+  implicit val compliancePercentEncoder: JsonEncoder[SystemUpdateStats] = DeriveJsonEncoder.gen
+
+  override def parse(score: Score[Json]): IOResult[Option[SystemUpdateScore]] = {
+
+    if (score.name == "system-updates") {
+      score.details.as.toIO.map(SystemUpdateScore(score.value, score.message, _)).map(Some(_))
+    } else {
+      None.succeed
+    }
+  }
+
+  override def toJson(score: Score[_]): IOResult[Option[Score[Json]]] = {
+    score.details match {
+      case details: SystemUpdateStats =>
+        details.toJsonAST.toIO.map(JsonScore(score.value, score.name, score.message, _)).map(Some(_))
+      case _ => None.succeed
+    }
+  }
+}
+
+class ScoreSerializer {
+
+  val serializers: Ref[List[ScoreSerializerService]] =
+    Ref.make(ComplianceScoreSerializerService :: SystemUpdateScoreSerializerService :: List.empty[ScoreSerializerService]).runNow
+
+  def registerSerializer(handler: ScoreSerializerService) = {
+    serializers.update(handler :: _)
+  }
+
+  def parse(score: Score[Json]): IOResult[Score[_]] = {
```

Comment:
I looks like the things are done so that several serializer can work on a Score (because of the logic of "try all serializer and get the first" vs "find a serializer that can handle that case and make it do its work"). Is there a case where it would be interesting to have several serializer succeeding or is this just no specific reasons ?

Reply:
The third commit in https://github.com/VinceMacBuche/rudder/pull/29 there is an example of using `collectFirst` with view to stop at the first matching one